### ISSUE TITLE
Add support for mixture models in torch.distributions

### ISSUE THREAD

**Issue by SkafteNicki** [pytorch/pytorch#18125]
## 🚀 Feature
It would be useful to have support for mixture models in torch.distributions e.g. Gaussian Mixture models. 

## Motivation

Currently, there are no easy way of defining mixture models/distributions in pytorch. I propose an re implementation of tensorflows api for doing this

```import tensorflow as tf
import tensorflow_probability as tfp
tfd = tfp.distributions

loc = tf.get_variable(name="loc", shape=[mixture_components, latent_size])
raw_scale_diag = tf.get_variable(name="scale", shape=[mixture_components, latent_size])
mixture_logits = tf.get_variable(name="mixture_logits", shape=[mixture_components])
mix = tfd.Categorical(logits=mixture_logits)
comp = tfd.MultivariateNormalDiag(loc=loc, scale_diag=tf.nn.softplus(raw_scale_diag))
gmm =  tfd.MixtureSameFamily(components_distribution=comp,
                             mixture_distribution=mix,
                             name="prior")
# Usefull methods
gmm.mean()
gmm.sample(...)
gmm.log_prob(...)
```

I personally use this for variational autoencoders, to be able to work with more expressive priors than the standard gaussian prior.

## Pitch

Re implementation of the [MixtureSameFamily](https://www.tensorflow.org/probability/api_docs/python/tfp/distributions/MixtureSameFamily) or [Mixture](tfp.distributions.Mixture) class from tensorflow. Personally, I consider the MixtureSameFamily class to be the most important of the two, because mixtures of different distributions is rarely considered in practice.

**Comment by Protocal13** [comment]
So whats the problem with it exactly

**Comment by Protocal13** [comment]
Please send code and the result so then i get a better understanding

**Comment by SsnL** [comment]
I agree that we should support mixture distribution. It shouldn't be that hard to do.

**Comment by apaszke** [comment]
cc @fritzo @neerajprad

**Comment by fritzo** [comment]
`MixtureSameFamily` should be easy to implement. We don't use this in Pyro since we usually keep the mixture component id as an explicit variable and enumerate over that variable:
```py
component = pyro.sample("component", dist.Categorical(probs),
                        infer={"enumerate": "parallel"})
assert component.reshape(-1).shape == probs.shape[-1:]
value = pyro.sample("mixture", MyDistribution(my_params[component]))
```
cc @martinjankowiak

**Comment by SkafteNicki** [comment]
I also agree that it should somewhat be easy to implement. Only complicated part would be the `rsample()` method which require a implicit reparameterization of the gradients to be able to backprop through ([Stochastic Backpropagation through Mixture Density Distributions, Alex Graves](https://arxiv.org/abs/1607.05690))

**Comment by hamfry** [comment]
Thanks for raising this.

Has anybody come up with an implementation of this (mixture of multivariate Gaussians)?

Would be very useful for loss function calculations.

**Comment by jhmenke** [comment]
> Thanks for raising this.
> 
> Has anybody come up with an implementation of this (mixture of multivariate Gaussians)?
> 
> Would be very useful for loss function calculations.

This is somewhat related: https://github.com/sagelywizard/pytorch-mdn

**Comment by ElleryL** [comment]
Hey I have implemented a mixture of Gaussian. The class do the following 

```
    m1 = MultivariateNormal(torch.zeros(2) + 10.,torch.eye(2) * 1)
    m2 = MultivariateNormal(torch.zeros(2) - 10.,torch.eye(2) * 1)
    mix = GMM(cat = Tensor([.5,.5]), components = [m1,m2], dim=2)
```

which creates a bivariate. Can I push this code ?

**Comment by fritzo** [comment]
If we add a mixture distribution, then I think it should be a generic `MixtureSameFamily`.

What are the use cases for such a distribution?

As mentioned above:
- `MixtureSameFamily` is not pathwise differentiable, so `.rsample()` cannot be implemented.
- Pyro avoids `MixtureSameFamily`, instead enumerating over components. This provides equivalent model flexibility, avoids the need for pathwise derivatives, yet uses each component's `.rsample()` method if available.

I believe the `Distribution` interface is not an appropriate place for sample-differentiable mixture models. Is a differentiable mixture model `.log_prob()` method useful?

**Comment by SkafteNicki** [comment]
I went ahead a implemented a ```MixtureSameFamily``` class using the ```Distribution``` interface. It is basically a copy of the tensorflow implementation which includes the ```mean, variance, sample, log_prob``` methods. The ```.rsample()``` method is still missing. Code can be seen here:

https://gist.github.com/SkafteNicki/4d582f0952941a98f920af6cb5f680a2

I can create i PR if wanted.

**Comment by fritzo** [comment]
@SkafteNicki sure, if you want to put up a PR I can review to suggest extra things needed for torch.distributions, like docs, tests, `.expand()` and metadata attributes. Could you cc me and @neerajprad ? Thanks!

**Comment by RylanSchaeffer** [comment]
What's the status of this issue? Any idea when the PR will be merged?

**Comment by YazidJanati** [comment]
I've coded the sampling function for the gaussian mixture model : 

```
def GM_samples(N_samples,weights,means,covs):
    dim, N_mixt = means.shape[1], weights.shape[1]
    categories  = OHC(weights).sample([N_samples])
    categories  = categories.view(N_samples,N_mixt,1,1)
    indic_covs  = (categories * covs).float()
    std_samples = MVN(tensor([0.]*dim),torch.eye(dim)).sample([N_samples])
    std_samples = std_samples.view(N_samples,1,dim,1)
    indic_means = (categories*(means.view(1,N_mixt,dim,1))).float()
    samples     = indic_means + indic_covs @ std_samples
    samples     = samples.view(N_samples,N_mixt,dim)
    ones        = tensor([[1.]*N_mixt])
    return (ones @ samples).view(N_samples,dim)
```

where :  
- weights is of shape (1,number of mixtures)
- means is of shape (number of mixtures, dim)
- covs is of shape (number of mixtures, dim, dim) 

I guess the code is pretty fast : 10k samples in 7.4 ms

**Comment by Svito-zar** [comment]
Looking forward to see this integrated in PyTorch!

**Comment by JTaets** [comment]
Is there still a plan to implement the rsample for mixture models (https://arxiv.org/abs/1607.05690)?
Tensorflow has it implemented and I think it could be usefull.(https://github.com/tensorflow/probability/blob/9415ba638bd3b372512241c54d254eb076c566df/tensorflow_probability/python/distributions/mixture_same_family.py#L433)

**Comment by fritzo** [comment]
cc @martinjankowiak (re: `MixtureSameFamily.rsample()`)

**Comment by CXX1113** [comment]
> 

Can this implementation backpropagate gradients?

**Comment by CXX1113** [comment]
> I agree that having GMM implemented would be of a great benefit.
> 
> I've coded the sampling function for the gaussian mixture model :
> 
> ```
> def GM_samples(N_samples,weights,means,covs):
>     dim, N_mixt = means.shape[1], weights.shape[1]
>     categories  = OHC(weights).sample([N_samples])
>     categories  = categories.view(N_samples,N_mixt,1,1)
>     indic_covs  = (categories * covs).float()
>     std_samples = MVN(tensor([0.]*dim),torch.eye(dim)).sample([N_samples])
>     std_samples = std_samples.view(N_samples,1,dim,1)
>     indic_means = (categories*(means.view(1,N_mixt,dim,1))).float()
>     samples     = indic_means + indic_covs @ std_samples
>     samples     = samples.view(N_samples,N_mixt,dim)
>     ones        = tensor([[1.]*N_mixt])
>     return (ones @ samples).view(N_samples,dim)
> ```
> 
> where :
> 
> * weights is of shape (1,number of mixtures)
> * means is of shape (number of mixtures, dim)
> * covs is of shape (number of mixtures, dim, dim)
> 
> (by cov i mean the upper triangular matrix in the cholesky decomposition of the covariance matrix) I guess the code is pretty fast : 10k samples in 7.4 ms
> 
> For example :
> 
> ```
> weights = tensor([[0.3,0.7]])
> means   = tensor([[0,1],[2,3]])
> covs    = tensor([np.eye(2),2*np.eye(2)])
> 
> %time samples = GM_samples(10000,weights,means,covs)
> ```


Can this implementation backpropagate gradients?
